### PR TITLE
Report descending sort option

### LINKED ISSUE #199: Add a way to sort the text report
Hi Ned,

 

The HTML output looks very good, however I’m running the coverage tool on a VMware and just want to mail the output as plain text to my colleagues.

Sending all those HTML files to all of them (or copy them to a server) is too much, so I need another trick - maybe just a python filter that sorts it out.

 

Thanks for your time!

Rob

 

From: Ned Batchelder [mailto:ned@nedbatchelder.com]
Sent: Friday, September 28, 2012 14:56
To: Tillaart, Rob van den
Subject: Re: question about coverage tool

 

Hi Rob,

In the HTML report, you can click on the column headers to sort by the column you're interested in.

--Ned.

On 9/28/2012 7:33 AM, Tillaart, Rob van den wrote:

    Hi Ned,

     

    Just new to python development I found the coverage tool which does a good job for Q&A.

     

    Q: is there an option to sort the output on coverage %?

    If not I would like to put it on the “wishlist” for next release.

     

    A sorted list makes it easier to see which files needs attention most,

     

    Thanks in advance,

    Rob Tillaart

    Netherlands

     

 

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/199

### PR DESCRIPTION
**Pull request by jerinpetergeorge**
Fix https://github.com/nedbat/coveragepy/issues/199

### CODE REVIEW

**Comment by jerinpetergeorge** [comment]
@nedbat Would you mind guide me to add the command line options? I am not much sure whether my commit, 0228c1a is okey or not for this.

**Comment by nedbat** [comment]
I've made you a pull request that fixes the tests: https://github.com/jerinpetergeorge/coveragepy/pull/1

File: coverage/cmdline.py [hunk 1]
```diff
@@ -85,6 +85,10 @@ class Opts(object):
             "which isn't done by default."
         ),
     )
+    sort = optparse.make_option(
+        '--sort', action='store', metavar='COLUMN',
+        help="Sort the report by the named column"
```
Comment (nedbat) [inline coverage/cmdline.py]: Let's add the choices here:
```suggestion
        help="Sort the report by the named column: name, stmts, miss, branch, brpart, or cover. Default is name."
```

File: coverage/summary.py [hunk 2]
```diff
@@ -104,10 +104,21 @@ def report(self, morfs, outfile=None):
...
+            _sort = self.config.sort.lower()
+            if _sort[0] == '-':
+                reverse = True
+                sort_option = _sort[1:]
+            elif _sort[0] == '+':
+                reverse = False
+                sort_option = _sort[1:]
+            else:
+                reverse = False
+                sort_option = _sort
```
Comment (nedbat) [inline coverage/summary.py]: We can tighten up this code a bit:
```suggestion
            sort_option = self.config.sort.lower()
            reverse = False
            if sort_option[0] == '-':
                reverse = True
                sort_option = sort_option[1:]
            elif sort_option[0] == '+':
                sort_option = sort_option[1:]
```

**Comment by nedbat** [comment]
BTW, at any time, you can say, "I've done as much as I want to do," and I'll pick it up, or we can keep working together on it.

File: coverage/summary.py (hunk 2 above)
Comment (jerinpetergeorge) [reply coverage/summary.py]: yeah, this is much better.

**Comment by nedbat** [comment]
Thanks! Are we good to merge?

**Comment by jerinpetergeorge** [comment]
I think we are good to go. I will make another PR with a few tests that probably cover the new sort option.

**Comment by nedbat** [comment]
We can wait for tests :)

**Comment by jerinpetergeorge** [comment]
Great. Let me try

**Comment by nedbat** [comment]
I'm thinking about a new release in a few days, and I'd like to get this into it.  I don't mind doing the tests myself if you don't mind.

**Comment by jerinpetergeorge** [comment]
> I'm thinking about a new release in a few days, and I'd like to get this into it. I don't mind doing the tests myself if you don't mind.

That's so fast... Anyway, I have pushed one more commit.

BTW, I came across this [piece of the test](https://github.com/nedbat/coveragepy/blob/97f05c988545f8ca588253fd325401c4834543b6/tests/test_summary.py#L909).
```
def test_sort_report_by_stmts(self):
    # Sort the text report by the Stmts column.
    report = self.get_summary_text(('report:sort', 'Stmts'))
    self.assert_ordering(report, "test_backward.py", "test_coverage.py", "test_api.py")
```

Just out of curiosity, I have changed the assert statement to something like `self.assert_ordering(report, "test_coverage.py", "test_backward.py", "test_api.py")` and ran the test. The tests are successfully executed.

Did I miss something here? @nedbat

**Comment by nedbat** [comment]
Wow, good catch on that broken test!  I've fixed it in f28b1dba5d8359029f5f491fec207c10097baf10, and added an assert to prevent that kind of breakage in the future.

Thanks for this, I'm merging it!

**Comment by jerinpetergeorge** [comment]
Welcome :smile: I hope you can handle the rest of the test cases if any

**Comment by nedbat** [comment]
This is now released as part of [coverage 5.2](https://pypi.org/project/coverage/5.2).

**Pull request merged**